server/src/main/java/org/elasticsearch/action/support/RefCountingRunnable.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.core.AbstractRefCounted;
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.Releasable;
+
+/**
+ * A mechanism to trigger an action on the completion of some (dynamic) collection of other actions. Basic usage is as follows:
+ *
+ * <pre>
+ * try (var refs = new RefCountingRunnable(finalRunnable)) {
+ *     for (var item : collection) {
+ *         runAsyncAction(item, refs.acquire()); // releases the acquired ref on completion
+ *     }
+ * }
+ * </pre>
+ *
+ * The delegate action is completed when execution leaves the try-with-resources block and every acquired reference is released. Unlike a
+ * {@link CountDown} there is no need to declare the number of subsidiary actions up front (refs can be acquired dynamically as needed) nor
+ * does the caller need to check for completion each time a reference is released. Moreover even outside the try-with-resources block you
+ * can continue to acquire additional listeners, even in a separate thread, as long as there's at least one listener outstanding:
+ *
+ * <pre>
+ * try (var refs = new RefCountingRunnable(finalRunnable)) {
+ *     for (var item : collection) {
+ *         if (condition(item)) {
+ *             runAsyncAction(item, refs.acquire());
+ *         }
+ *     }
+ *     if (flag) {
+ *         runOneOffAsyncAction(refs.acquire());
+ *         return;
+ *     }
+ *     for (var item : otherCollection) {
+ *         var itemRef = refs.acquire(); // delays completion while the background action is pending
+ *         executorService.execute(() -> {
+ *             try (var ignored = itemRef) {
+ *                 if (condition(item)) {
+ *                     runOtherAsyncAction(item, refs.acquire());
+ *                 }
+ *             }
+ *         });
+ *     }
+ * }
+ * </pre>
+ *
+ * In particular (and also unlike a {@link CountDown}) this works even if you don't acquire any extra refs at all: in that case, the
+ * delegate action executes at the end of the try-with-resources block.
+ */
+public final class RefCountingRunnable implements Releasable {
```

Comment:
`Releasable` isn't the right thing to use here, `Closeable` explicitly says the `close` is one-shot, and closes everything on the first call. But here it closes everything on the *last* call.

At the very least, this should be pointed out in the docs that `close` breaks contract.

Reply:
I think it's ok for `acquire` to return a separate `Releasable` lambda object that is one-shot, and that `RefCountedRunnable.close` is only called once by the try-with-resources, but having the same `close` method called multiple times feels wrong here.

eg

    public Releasable acquire() {
        return this::releaseRef;
    }

    private void releaseRef() {
        // decRef
    }

    public void close() {
        releaseRef();
    }

Reply:
That's a valid point, but note that we fairly extensively use other `Releasable` implementations that don't have idempotent `close()` methods too. Maybe `Releasable` should extend `AutoCloseable` directly?

Abusing this will typically trip the assertion in `AbstractRefCounted#decRef`. It would be possible to make each `Releasable` in play here a one-shot thing, but that would require more allocation and I don't think it really makes a difference.

I'll note this in the JavaDocs anyway.

Reply:
Here are some other examples of using `Releasable` in the same way:

- `org.elasticsearch.index.store.Store` (admittedly `close()` is idempotent here but the store lives on after it's been called, as long as there are other refs)
- `org.elasticsearch.common.bytes.ReleasableBytesReference` (exactly the same pattern as here)

Reply:
I think you're suggesting something like this?

```diff
diff --git a/server/src/main/java/org/elasticsearch/action/support/RefCountingRunnable.java b/server/src/main/java/org/elasticsearch/action/support/RefCountingRunnable.java
index 2667236d68f..0aa4cc4313d 100644
--- a/server/src/main/java/org/elasticsearch/action/support/RefCountingRunnable.java
+++ b/server/src/main/java/org/elasticsearch/action/support/RefCountingRunnable.java
@@ -16,6 +16,8 @@ import org.elasticsearch.core.AbstractRefCounted;
 import org.elasticsearch.core.RefCounted;
 import org.elasticsearch.core.Releasable;

+import java.util.concurrent.atomic.AtomicBoolean;
+
 /**
  * A mechanism to trigger an action on the completion of some (dynamic) collection of other actions. Basic usage is as follows:
  *
@@ -65,6 +67,20 @@ public final class RefCountingRunnable implements Releasable {
     static final String ALREADY_CLOSED_MESSAGE = "already closed, cannot acquire or release any further refs";

     private final RefCounted refCounted;
+    private final AtomicBoolean originalRefReleased = new AtomicBoolean();
+
+    private class AcquiredRef implements Releasable {
+        private final AtomicBoolean released = new AtomicBoolean();
+        @Override
+        public void close() {
+            releaseRef(released);
+        }
+
+        @Override
+        public String toString() {
+            return RefCountingRunnable.this.toString();
+        }
+    }

     /**
      * Construct a {@link RefCountingRunnable} which executes {@code delegate} when all refs are released.
@@ -83,7 +99,7 @@ public final class RefCountingRunnable implements Releasable {
     public Releasable acquire() {
         if (refCounted.tryIncRef()) {
             // closing ourselves releases a ref, so we can just return 'this' and avoid any allocation; callers only see a Releasable
-            return this;
+            return new AcquiredRef();
         }
         assert false : ALREADY_CLOSED_MESSAGE;
         throw new IllegalStateException(ALREADY_CLOSED_MESSAGE);
@@ -104,11 +120,19 @@ public final class RefCountingRunnable implements Releasable {
      */
     @Override
     public void close() {
-        try {
-            refCounted.decRef();
-        } catch (Exception e) {
-            logger.error("exception in delegate", e);
-            assert false : e;
+        releaseRef(originalRefReleased);
+    }
+
+    private void releaseRef(AtomicBoolean released) {
+        if (released.compareAndSet(false, true)) {
+            try {
+                refCounted.decRef();
+            } catch (Exception e) {
+                logger.error("exception in delegate", e);
+                assert false : e;
+            }
+        } else {
+            assert false : "already closed";
         }
     }
```

If so, I can live with that. It does seem a little unnecessary to do all those extra allocations but maybe I'm too optimistic that this won't be abused as you think :)

Reply:
Much better - I think with concurrency code the more explicit & error-checking it is, the better.

You could make the returned `AcquiredRef` idempotent in itself, so multiple closes don't throw exceptions, but also don't cause extra releases in the underlying object. Up to you

Reply:
> You could make the returned AcquiredRef idempotent in itself, so multiple closes don't throw exceptions, but also don't cause extra releases in the underlying object.

That's what it does now in production (i.e. with assertions disabled). In tests I think an unexpected double-close is a logic error; in the very few cases in which it might be permitted, the caller can be explicit about this by using `Releasables.releaseOnce(refs.acquire())`.

Reply:
Ah yes - didn't spot that method exists.